mmdet3d/core/bbox/coders/km3d_decoder.py

```diff
@@ -0,0 +1,413 @@
+    topk_xs = (topk_inds % width).int().float()
+
+    return topk_scores, topk_inds, topk_ys, topk_xs
+
+
+def _gather_feat(feat, ind, mask=None):
+    """Gather features with assigned index, with index expanded to the third
+    dimension.
+
+    Args:
+        feat (torch.Tensor): Features from model in shape (batch_size *
+            channels * height * width)
+        ind (torch.Tensor): Index of features to be gathered
+
+    Returns:
+        torch.Tensor: gathered features, in shape (batch_size * channels *
+        height * width)
+    """
+
+    # expand ind with the third dimension to be the same as feat
+    dim = feat.size(2)
+    ind = ind.unsqueeze(2).expand(ind.size(0), ind.size(1), dim)
+    feat = feat.gather(1, ind)
+    if mask is not None:
+        mask = mask.unsqueeze(2).expand_as(feat)
+        feat = feat[mask]
+        feat = feat.view(-1, dim)
+    return feat
+
+
+def _permute_and_gather_feat(feat, ind):
+    """Gather features by assigned index, with features permuted with (0,2,3,1)
+    and reshaped to three dimensions.
+
+    Args:
+        feat (torch.Tensor): Features from model in shape (batch_size *
+            channels * height * width)
+        ind (torch.Tensor): Index of features to be gathered
+
+    Returns:
+        torch.Tensor: Permuted and gathered features, in shape (batch_size *
+        channels * height * width)
+    """
+    feat = feat.permute(0, 2, 3, 1)
+    feat = feat.view(feat.size(0), -1, feat.size(3))
+    feat = _gather_feat(feat, ind)
+    return feat
+
+
+def _topk(scores, K=40):
+    """Select top k batches of scores, index, classes, y's, and x's given
+    scores.
+
+    Args:
+        scores (torch.Tensor): Scores from heatmap key points parameters, in
+            shape (batch_size * channels * height * width)
+        K (int): The number of top values to be output
+
+    Returns:
+        tuple[torch.Tensor]: Top-k scores, indices, classes, y and x
+        coordinates
+    """
+    batch, cat, height, width = scores.size()
+
+    topk_scores, topk_inds = torch.topk(scores.view(batch, cat, -1), K)
+
+    topk_inds = topk_inds % (height * width)
+    topk_ys = (topk_inds / width).int().float()
+    topk_xs = (topk_inds % width).int().float()
+
+    topk_score, topk_ind = torch.topk(topk_scores.view(batch, -1), K)
+    topk_clses = (topk_ind / K).int()
+    topk_inds = _gather_feat(topk_inds.view(batch, -1, 1),
+                             topk_ind).view(batch, K)
+    topk_ys = _gather_feat(topk_ys.view(batch, -1, 1), topk_ind).view(batch, K)
+    topk_xs = _gather_feat(topk_xs.view(batch, -1, 1), topk_ind).view(batch, K)
+
+    return topk_score, topk_inds, topk_clses, topk_ys, topk_xs
+
+
+def transform_img2cam(kps, dim, rot, meta, const):
+    """Transform coordinates in the 2D image plane to the 3D camera coordinate
+    system.
+
+    For detailed explanations of matrix, please refer to KM3D's paper:
+    https://arxiv.org/abs/2009.00764v1
```

Comment:
Please refer to this file for the docstring format with reference (at the [beginning](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/datasets/nuscenes_dataset.py#L17) and [url](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/datasets/nuscenes_dataset.py#L21)).